gh-90949: Recommend `hasattr` with Expat security methods

hartwork opened the pull request:
Same idea (and approach) as #116278.

CC @picnixz 

<!-- gh-issue-number: gh-90949 -->
* Issue: gh-90949
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--139800.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

hartwork committed:
Doc/library/pyexpat.rst: Recommend "hasattr" with Expat security methods

This mimics existing method SetReparseDeferralEnabled.

hartwork committed:
Doc/library/pyexpat.rst: Use ".. note::" with SetReparseDeferralEnabled

.. to make it consistent with the other four Expat security methods.

hartwork committed:
Doc/library/xml.etree.elementtree.rst: Use ".. note::" for backport notes

.. to make it consistent with Doc/library/pyexpat.rst.

StanFromIreland commented on Doc/library/xml.etree.elementtree.rst:
```diff
@@ -1398,7 +1398,9 @@ XMLParser Objects
       :meth:`xml.parsers.expat.xmlparser.SetReparseDeferralEnabled` for details.
 
-      Note that :meth:`flush` has been backported to some prior releases of
+      .. note::
```
This is causing failures, the text beneath it is not indented. However, I don't see a need for making it a note.

hartwork replied:
@StanFromIreland thanks for the review, fixed with the next push… (if indent by 3 is indeed the right fix)

StanFromIreland commented on Doc/library/xml.etree.elementtree.rst:
```diff
@@ -1476,7 +1478,9 @@ XMLPullParser Objects
       :meth:`xml.parsers.expat.xmlparser.SetReparseDeferralEnabled` for details.
 
-      Note that :meth:`flush` has been backported to some prior releases of
+      .. note::
```
Ditto.

hartwork committed:
Doc/library/xml.etree.elementtree.rst: Add missing indent by 3

picnixz commented on Doc/library/pyexpat.rst:
```diff
@@ -223,10 +223,14 @@ XMLParser Objects
+      has been backported to some prior releases of CPython as a security fix.
+      Check for availability of
+      :meth:`SetReparseDeferralEnabled`
+      using :func:`hasattr` if used in code running across a variety of Python
```
```suggestion
      :meth:`!SetReparseDeferralEnabled`
      has been backported to some prior releases of CPython as a security fix.
      Check for availability of
      :meth:`!SetReparseDeferralEnabled`
      using :func:`hasattr` if used in code running across a variety of Python
```

We usually avoid putting a link to the object being currently documented (https://devguide.python.org/documentation/style-guide/#links)

A simpler way to do it is to say "This method" instead of restating the name.

hartwork replied:
@picnixz I would like to note that:

- the linked docs also state "It’s natural to use the name of a function in the documentation for the function" and that
- the distance to the name of the section (and hence the function talked about) is more than just a few lines (or centimeters on screen) and that distance may change the picture a bit.

~Can we agree on using the name but without a link (but two backticks before and after)?~

hartwork replied:
@picnixz PS: I'm only noticing the `!` there now. Let me do that, one minute…

hartwork replied:
@picnixz updated :+1:

hartwork committed:
Add an exclemation mark to suppress distracting links

Related:
https://devguide.python.org/documentation/style-guide/#links

Idea by @picnixz

hartwork committed:
Demo more quiet version without ".. note::"

hugovk commented on Doc/library/pyexpat.rst:
```diff
@@ -257,11 +259,19 @@ against some common XML vulnerabilities.
       Activation thresholds below 4 MiB are known to break support for DITA 1.3
       payload and are hence not recommended.
 
+
```
```suggestion
```

hartwork replied:
@hugovk my bad, I missed this because of using `ydiff`. Fixed with the next push…

hugovk commented on Doc/library/pyexpat.rst:
```diff
@@ -223,10 +223,12 @@ XMLParser Objects
+   Check for availability of
+   :meth:`!SetReparseDeferralEnabled`
+   using :func:`hasattr` if used in code running across a variety of Python
+   versions.
```
Suggestion to make a bit shorter: we don't necessarily need to say "Note that" or repeat the method name.

```suggestion
   This method has been backported to some earlier releases of CPython as a
   security fix. Check for its availability using :func:`hasattr` for code
   running across a number of Python versions.
```

Can apply to the others too.

And rather than "backported to some earlier releases of CPython", can we say specify the oldest version? "backported to CPython 3.[x]"?

hartwork replied:
> Suggestion to make a bit shorter: we don't necessarily need to say "Note that" or repeat the method name.

@hugovk I don't fully agree but will follow your suggestion. Upcoming…

> Can apply to the others too.

Sure!

> And rather than "backported to some earlier releases of CPython", can we say specify the oldest version? "backported to CPython 3.[x]"?

That would be trouble, e.g. it would mis-suggest that if the user is running a version greater or equal to that number that they will have that method available which is not the case… Please note that that part has passed review in the past in #116278. I believe there is no bug to fix there but rather a risk of making things worse: misleading or incorrect. Please vote for keeping that aspect as-is.

hugovk replied:
Yes, that's fine 👍

hartwork replied:
@hugovk thank you! :+1:

hartwork committed:
Revert accidental addition of blank line

hartwork committed:
Drop "Note that " at the start of the sentence

Requested by @hugovk

hartwork committed:
Resolve double self reference

As requested by @hugovk

hugovk reviewed (approved):
Thank you!
